// @databases/pg-migrations/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.restoreMigrationFromDatabase = exports.markMigrationAsUnapplied = exports.markMigrationAsApplied = exports.ignoreError = exports.applyMigrations = exports.sql = exports.connect = void 0;

const migrations_base_1 = require("@databases/migrations-base");

const pg_1 = require("@databases/pg");

exports.connect = pg_1.default;
Object.defineProperty(exports, "sql", {
  enumerable: true,
  get: function () {
    return pg_1.sql;
  }
});

const PostgresDatabaseEngine_1 = require("./PostgresDatabaseEngine");

const assertIsDirectory_1 = require("./assertIsDirectory");

const {
  applyMigrations,
  ignoreError,
  markMigrationAsApplied,
  markMigrationAsUnapplied,
  restoreMigrationFromDatabase
} = migrations_base_1.getPublicApi(({
  connection,
  migrationsDirectory,
  versionTableName = 'atdatabases_migrations_version',
  appliedMigrationsTableName = 'atdatabases_migrations_applied'
}) => new PostgresDatabaseEngine_1.default(connection, {
  versionTableName,
  appliedMigrationsTableName,
  migrationsDirectory: assertIsDirectory_1.default(migrationsDirectory, 'throw')
}));
exports.applyMigrations = applyMigrations;
exports.ignoreError = ignoreError;
exports.markMigrationAsApplied = markMigrationAsApplied;
exports.markMigrationAsUnapplied = markMigrationAsUnapplied;
exports.restoreMigrationFromDatabase = restoreMigrationFromDatabase;
