// @databases/migrations-base/lib/methods/readMigrationsDir.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const Result_1 = require("../types/Result");

function notNull(value) {
  return value != null;
}

const IGNORED_EXTENSIONS = ['.d.ts'];
const INCLUDED_EXTENSIONS = ['.sql', '.js', '.mjs', '.jsx', '.ts', '.tsx'];

async function readMigrationsDir(migrationsDirectory) {
  const invalidFilenames = [];
  const migrations = (await Promise.all((await migrationsDirectory.listFiles()).map(async fileName => {
    if (fileName[0] === '_') return null;

    if (IGNORED_EXTENSIONS.some(e => fileName.endsWith(e)) || !INCLUDED_EXTENSIONS.some(e => fileName.endsWith(e))) {
      return null;
    }

    const match = /^(\d+)\-/.exec(fileName);

    if (!match) {
      invalidFilenames.push(fileName);
      return null;
    }

    const index = parseInt(match[1], 10);
    const src = await migrationsDirectory.read(fileName);
    return {
      index,
      name: fileName,
      script: src
    };
  }))).filter(notNull);

  if (invalidFilenames.length) {
    return Result_1.default.fail({
      code: 'migration_filenames',
      files: invalidFilenames
    });
  }

  return Result_1.default.ok(migrations);
}

exports.default = readMigrationsDir;
