// @databases/migrations-base/lib/DirectoryContext.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const fs_1 = require("fs");

const path_1 = require("path");

function notNull(value) {
  return value != null;
}

class DirectoryContext {
  constructor(directory, loadMigration) {
    this._directory = directory;
    this._loadMigration = loadMigration;
  }

  _resolve(filename) {
    return path_1.join(this._directory, filename);
  }

  async listFiles() {
    return (await Promise.all((await fs_1.promises.readdir(this._directory)).map(async file => {
      return (await fs_1.promises.stat(this._resolve(file))).isFile() ? file : null;
    }))).filter(notNull);
  }

  async read(filename) {
    return await fs_1.promises.readFile(this._resolve(filename), 'utf8');
  }

  async write(filename, content) {
    await fs_1.promises.writeFile(this._resolve(filename), content);
  }

  async rename(fromFilename, toFilename) {
    await fs_1.promises.rename(this._resolve(fromFilename), this._resolve(toFilename));
  }

  async delete(filename) {
    await fs_1.promises.unlink(this._resolve(filename));
  }

  loadMigration(filename) {
    return this._loadMigration(this._resolve(filename));
  }

}

exports.default = DirectoryContext;
