// @databases/migrations-base/lib/getPublicApi.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const runCommand_1 = require("./runCommand");

const applyMigrations_1 = require("./commands/applyMigrations");

const throwError_1 = require("./methods/throwError");

const ignoreError_1 = require("./commands/ignoreError");

const markMigrationAsApplied_1 = require("./commands/markMigrationAsApplied");

const markMigrationAsUnapplied_1 = require("./commands/markMigrationAsUnapplied");

const restoreMigrationFromDatabase_1 = require("./commands/restoreMigrationFromDatabase");

function getPublicApi(getEngine) {
  async function run(runParameters, commandParameters, command) {
    const engine = getEngine(runParameters);
    const result = await runCommand_1.default(engine, command, commandParameters, runParameters);

    if (!result.ok) {
      throwError_1.default(result.reason, engine);
    }
  }

  return {
    async applyMigrations(parameters) {
      await run(parameters, {
        ignored_errors: parameters.ignoredErrors
      }, applyMigrations_1.default());
    },

    async ignoreError(parameters) {
      await run(parameters, {
        error_type: parameters.errorType,
        applied_migration: parameters.migrationIndex
      }, ignoreError_1.default());
    },

    async markMigrationAsApplied(parameters) {
      await run(parameters, {
        migration_file: parameters.migrationIndex
      }, markMigrationAsApplied_1.default());
    },

    async markMigrationAsUnapplied(parameters) {
      await run(parameters, {
        applied_migration: parameters.migrationIndex
      }, markMigrationAsUnapplied_1.default());
    },

    async restoreMigrationFromDatabase(parameters) {
      await run(parameters, {
        applied_migration: parameters.migrationIndex
      }, restoreMigrationFromDatabase_1.default());
    }

  };
}

exports.default = getPublicApi;
