// @databases/migrations-base/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var Result_1 = require("./types/Result");

exports.Result = Result_1.default;

var DirectoryContext_1 = require("./DirectoryContext");

exports.DirectoryContext = DirectoryContext_1.default;

var MigrationContext_1 = require("./MigrationContext");

exports.MigrationContext = MigrationContext_1.default;

var runCommand_1 = require("./runCommand");

exports.runCommand = runCommand_1.default;

var applyMigrations_1 = require("./commands/applyMigrations");

exports.applyMigrations = applyMigrations_1.default;

var ignoreError_1 = require("./commands/ignoreError");

exports.ignoreError = ignoreError_1.default;

var markMigrationAsApplied_1 = require("./commands/markMigrationAsApplied");

exports.markMigrationAsApplied = markMigrationAsApplied_1.default;

var markMigrationAsUnapplied_1 = require("./commands/markMigrationAsUnapplied");

exports.markMigrationAsUnapplied = markMigrationAsUnapplied_1.default;

var restoreMigrationFromDatabase_1 = require("./commands/restoreMigrationFromDatabase");

exports.restoreMigrationFromDatabase = restoreMigrationFromDatabase_1.default;

var getPublicApi_1 = require("./getPublicApi");

exports.getPublicApi = getPublicApi_1.default;

var getCommandLineInterface_1 = require("./getCommandLineInterface");

exports.getCommandLineInterface = getCommandLineInterface_1.default;
exports.commands = getCommandLineInterface_1.commands;

var printError_1 = require("./methods/printError");

exports.printError = printError_1.default;
exports.parameters = require("parameter-reducers");
